File: ACTIVIDADES/Actividad N°3/ListaEnlazadaSimple.cpp
#include <iostream>
using namespace std;

// Definición del nodo
struct Nodo {
    int dato;
    Nodo* siguiente;
    
    Nodo(int valor) : dato(valor), siguiente(nullptr) {}
};

class ListaEnlazada {
private:
    Nodo* cabeza;
    
public:
    ListaEnlazada() : cabeza(nullptr) {}
    
    // Destructor para liberar memoria
    ~ListaEnlazada() {
        while (cabeza != nullptr) {
            Nodo* temp = cabeza;
            cabeza = cabeza->siguiente;
            delete temp;
        }
    }
    
    // Insertar al inicio
    void insertarAlInicio(int valor) {
        Nodo* nuevoNodo = new Nodo(valor);
        nuevoNodo->siguiente = cabeza;
        cabeza = nuevoNodo;
        cout << "Insertado al inicio: " << valor << endl;
    }
    
    // Insertar al final
    void insertarAlFinal(int valor) {
        Nodo* nuevoNodo = new Nodo(valor);
        
        if (cabeza == nullptr) {
            cabeza = nuevoNodo;
        } else {
            Nodo* actual = cabeza;
            while (actual->siguiente != nullptr) {
                actual = actual->siguiente;
            }
            actual->siguiente = nuevoNodo;
        }
        cout << "Insertado al final: " << valor << endl;
    }
    
    // Eliminar del inicio
    int eliminarDelInicio() {
        if (cabeza == nullptr) {
            cout << "Error: Lista vacia" << endl;
            return -1;
        }
        
        Nodo* temp = cabeza;
        int valor = temp->dato;
        cabeza = cabeza->siguiente;
        delete temp;
        cout << "Eliminado del inicio: " << valor << endl;
        return valor;
    }
    
    // Mostrar lista
    void mostrar() {
        if (cabeza == nullptr) {
            cout << "Lista vacia" << endl;
            return;
        }
        
        cout << "Recorrido completo: ";
        Nodo* actual = cabeza;
        while (actual != nullptr) {
            cout << actual->dato;
            if (actual->siguiente != nullptr) {
                cout << " -> ";
            }
            actual = actual->siguiente;
        }
        cout << " -> NULL" << endl;
    }
    
    bool estaVacia() {
        return cabeza == nullptr;
    }
};

// Función principal
int main() {
    ListaEnlazada lista;
    
    cout << "=== IMPLEMENTACION DE LISTA ENLAZADA EN C++ ===" << endl;
    
    // Insertar al inicio los valores: 8, 4
    cout << "\n1. Insertando al inicio:" << endl;
    lista.insertarAlInicio(8);
    lista.insertarAlInicio(4);
    
    cout << "\nEstado despues de insertar al inicio:" << endl;
    lista.mostrar();
    
    // Insertar al final el valor: 11
    cout << "\n2. Insertando al final:" << endl;
    lista.insertarAlFinal(11);
    
    cout << "\nEstado despues de insertar al final:" << endl;
    lista.mostrar();
    
    // Eliminar el primer nodo
    cout << "\n3. Eliminando el primer nodo:" << endl;
    lista.eliminarDelInicio();
    
    // Mostrar el recorrido completo
    cout << "\n4. Recorrido final:" << endl;
    lista.mostrar();
    
    return 0;
}

File: ACTIVIDADES/Actividad N°3/ListaEnlazadaSimple.py
class Nodo:
    def __init__(self, dato):
        self.dato = dato
        self.siguiente = None

class ListaEnlazada:
    def __init__(self):
        self.cabeza = None
    
    def insertar_al_inicio(self, valor):
        """Insertar al inicio"""
        nuevo_nodo = Nodo(valor)
        nuevo_nodo.siguiente = self.cabeza
        self.cabeza = nuevo_nodo
        print(f"Insertado al inicio: {valor}")
    
    def insertar_al_final(self, valor):
        """Insertar al final"""
        nuevo_nodo = Nodo(valor)
        
        if self.cabeza is None:
            self.cabeza = nuevo_nodo
        else:
            actual = self.cabeza
            while actual.siguiente is not None:
                actual = actual.siguiente
            actual.siguiente = nuevo_nodo
        print(f"Insertado al final: {valor}")
    
    def eliminar_del_inicio(self):
        """Eliminar del inicio"""
        if self.cabeza is None:
            print("Error: Lista vacía")
            return None
        
        valor = self.cabeza.dato
        self.cabeza = self.cabeza.siguiente
        print(f"Eliminado del inicio: {valor}")
        return valor
    
    def mostrar(self):
        """Mostrar lista"""
        if self.cabeza is None:
            print("Lista vacía")
            return
        
        print("Recorrido completo: ", end="")
        actual = self.cabeza
        while actual is not None:
            print(actual.dato, end="")
            if actual.siguiente is not None:
                print(" -> ", end="")
            actual = actual.siguiente
        print(" -> NULL")
    
    def esta_vacia(self):
        return self.cabeza is None

# Función principal
def main():
    lista = ListaEnlazada()
    
    print("=== IMPLEMENTACIÓN DE LISTA ENLAZADA EN PYTHON ===")
    
    # Insertar al inicio los valores: 8, 4
    print("\n1. Insertando al inicio:")
    lista.insertar_al_inicio(8)
    lista.insertar_al_inicio(4)
    
    print("\nEstado después de insertar al inicio:")
    lista.mostrar()
    
    # Insertar al final el valor: 11
    print("\n2. Insertando al final:")
    lista.insertar_al_final(11)
    
    print("\nEstado después de insertar al final:")
    lista.mostrar()
    
    # Eliminar el primer nodo
    print("\n3. Eliminando el primer nodo:")
    lista.eliminar_del_inicio()
    
    # Mostrar el recorrido completo
    print("\n4. Recorrido final:")
    lista.mostrar()

if __name__ == "__main__":
    main()
